skorch/net.py

```diff
@@ -243,11 +236,33 @@ class NeuralNet:
       The complete (i.e. default and other), initialized callbacks, in
       a tuple with unique names.
 
+    modules_ : list of str
+      List of names of all modules that are torch modules. This list is
+      collected dynamically when the net is initialized. Typically, there is no
+      reason for a user to modify this list.
+
+    criteria_ : list of str
+      List of names of all criteria that are torch modules. This list is
+      collected dynamically when the net is initialized. Typically, there is no
+      reason for a user to modify this list.
+
+    optimizers_ : list of str
+      List of names of all optimizers. This list is collected dynamically when
+      the net is initialized. Typically, there is no reason for a user to modify
+      this list.
+
     """
-    prefixes_ = ['module', 'iterator_train', 'iterator_valid', 'optimizer',
-                 'criterion', 'callbacks', 'dataset']
+    prefixes_ = ['iterator_train', 'iterator_valid', 'callbacks', 'dataset']
+
+    cuda_dependent_attributes_ = []
 
-    cuda_dependent_attributes_ = ['module_', 'optimizer_', 'criterion_']
+    # This attribute keeps track of which initialization method is being used.
+    # It should not be changed manually.
+    init_context_ = None
+
+    modules_ = []
+    criteria_ = []
+    optimizers_ = []
```

Comment:
Typically, this would not work with grid search, but the way these attributes are updated with `__setattr__` allows it to work.

Reply:
I agree it's not the most elegant of solutions, but I couldn't come up with anything better.